pkg/sources/s3/s3.go

```diff
@@ -132,79 +134,212 @@ func (s *Source) newClient(region string) (*s3.S3, error) {
-				s.log.Error(err, "could not make regional s3 client")
-			}
-		} else {
-			regionalClient = client
-		}
+	cfg.Credentials = baseCredentials
 
-		errorCount := sync.Map{}
+	sess, err = session.NewSessionWithOptions(session.Options{
+		SharedConfigState: session.SharedConfigEnable,
+		Config: *cfg,
+	})
 
-		err = regionalClient.ListObjectsV2PagesWithContext(
-			ctx, &s3.ListObjectsV2Input{Bucket: &bucket},
-			func(page *s3.ListObjectsV2Output, last bool) bool {
-				s.pageChunker(ctx, regionalClient, chunksChan, bucket, page, &errorCount, i+1, &objectCount)
-				return true
-			})
+	if err != nil {
+		return nil, err
+	}
+
+	// Create and return the S3 client with the final configuration
+	return s3.New(sess), nil
+}
 
-		if err != nil {
-			return fmt.Errorf(
-				"could not list objects in s3 bucket: bucket %s: %w",
-				bucket,
-				err)
+// Chunks emits chunks of bytes over a channel.
+func (s *Source) Chunks(ctx context.Context, chunksChan chan *sources.Chunk) error {
+	const defaultAWSRegion = "us-east-1"
+
+	if len(s.conn.Roles) > 0 {
+		for _, role := range s.conn.Roles {
+			
+		        client, err := s.newRoleClient(defaultAWSRegion, role)
+		        if err != nil {
+		        	return errors.WrapPrefix(err, "could not create s3 client", 0)
+		        }
+
+		        var bucketsToScan []string
+
+		        switch s.conn.GetCredential().(type) {
+		        case *sourcespb.S3_AccessKey, *sourcespb.S3_SessionToken, *sourcespb.S3_CloudEnvironment:
+		        	if len(s.conn.Buckets) == 0 {
+		        		res, err := client.ListBuckets(&s3.ListBucketsInput{})
+		        		if err != nil {
+		        			return fmt.Errorf("could not list s3 buckets: %w", err)
+		        		}
+		        		buckets := res.Buckets
+		        		for _, bucket := range buckets {
+		        			bucketsToScan = append(bucketsToScan, *bucket.Name)
+		        		}
+		        	} else {
+		        		bucketsToScan = s.conn.Buckets
+		        	}
+		        case *sourcespb.S3_Unauthenticated:
+		        	if len(s.conn.Buckets) == 0 {
+		        		res, err := client.ListBuckets(&s3.ListBucketsInput{})
+		        		if err != nil {
+		        			return fmt.Errorf("could not list s3 buckets: %w", err)
+		        		}
+		        		buckets := res.Buckets
+		        		for _, bucket := range buckets {
+		        			bucketsToScan = append(bucketsToScan, *bucket.Name)
+		        		}
+		        	} else {
+		        		bucketsToScan = s.conn.Buckets
+				}
+		        default:
+		        	return errors.Errorf("invalid configuration given for %s source", s.name)
+		        }
+
+		        objectCount := uint64(0)
+		        for i, bucket := range bucketsToScan {
+		        	if common.IsDone(ctx) {
+		        		return nil
+		        	}
+
+		        	s.SetProgressComplete(i, len(bucketsToScan), fmt.Sprintf("Bucket: %s", bucket), "")
+
+		        	s.log.Info("Scanning bucket", "bucket", bucket)
+		        	region, err := s3manager.GetBucketRegionWithClient(context.Background(), client, bucket)
```

Comment:
**Nit:** We can use the passed in `ctx` instead of making a `context.Background()`.